File: cpp/MatchableString/MatchableString/suffixarray/matchablestringsa.cpp
#include "matchablestringsa.h"

MatchableStringSA::MatchableStringSA(const char *orig) : MatchableString(orig) {
	substringArray = new int[length];
	for (int i = 0; i < length; i++)
	{
		substringArray[i] = i;
	}

	isPreProcessed = false;
}

MatchableStringSA::~MatchableStringSA() {
	delete[] substringArray;
}

void MatchableStringSA::preprocess(bool parallel) {
	auto comp = [this](int one, int other) {
		return strcmp(str + one, str + other) < 0;
	};

	if (!parallel) {
		std::sort(substringArray, substringArray + length, comp);
	}
	else {
		MergeSort<int>::parallelSort(substringArray, length, comp);
	}

	isPreProcessed = true;
}

int MatchableStringSA::binarySearch(int min, int max, const std::unique_ptr<ComparerBase>& comparer) const {
	if (min == max) {
		return -1;
	}

	int num = (min + max) / 2;

	switch (comparer->compare(num)) {
	case 0:
		return num;
	case -1:
		return binarySearch(num + 1, max, comparer);
	case 1:
		return binarySearch(min, num, comparer);
	default:
		throw std::logic_error("Not possible to get here - theoretically.");
	}
}

std::vector<int> MatchableStringSA::match(Pattern& pattern, bool parallel) const {
	if (!isPreProcessed) {
		throw std::runtime_error("This MatchAbleString hasn't been preprocessed yet!");
	}

	std::unique_ptr<ComparerBase> foc = std::make_unique<FirstOccurenceComparer>(str, pattern.getString(), substringArray, length);
	std::unique_ptr<ComparerBase> loc = std::make_unique<LastOccurenceComparer>(str, pattern.getString(), substringArray, length);

	int first = binarySearch(0, length, foc);
	int last = binarySearch(0, length, loc);

	if (first == -1)
	{
		return std::vector<int>();
	}
	else
	{
		std::vector<int> tmp;

		for (int i = first; i <= last; i++) {
			tmp.push_back(substringArray[i]);
		}

		std::sort(tmp.begin(), tmp.end());

		return tmp;
	}
}

File: go/MatchableString/matchablestringsa.go
package main

import "sort"

type MatchableStringSA struct {
	str            string
	length         int
	substringArray []string
	isPreProcessed bool
}

func (s *MatchableStringSA) SetString(str string) {
	s.str = str
	s.length = len(str)

	s.substringArray = make([]string, s.length)
	for i := 0; i < s.length; i++ {
		s.substringArray[i] = str[i:]
	}

	s.isPreProcessed = false
}

func (s *MatchableStringSA) Preprocess(parallel bool) {
	if !parallel {
		sort.Strings(s.substringArray)
	} else {
		acc := make([]string, s.length)

		parallelSort(s.substringArray, acc, false, nil)
	}

	s.isPreProcessed = true
}

func binarySearch(min int, max int, substringArray []string, pattern string, comparer func(int, []string, string) int) int {
	if min == max {
		return -1
	}

	var num int
	num = (min + max) / 2

	switch comparer(num, substringArray, pattern) {
	case 0:
		return num
	case -1:
		return binarySearch(num+1, max, substringArray, pattern, comparer)
	case 1:
		return binarySearch(min, num, substringArray, pattern, comparer)
	default:
		// No way we ever get here.
		panic("Unhandled error.")
	}
}

func (s *MatchableStringSA) Match(pattern *Pattern, parallel bool) []int {
	if !s.isPreProcessed {
		panic("This MatchAbleString hasn't been preprocessed yet!")
	}

	first := binarySearch(0, s.length, s.substringArray, pattern.str, firstOccurrenceComparer)
	last := binarySearch(0, s.length, s.substringArray, pattern.str, lastOccurrenceComparer)

	if first == -1 {
		return make([]int, 0)
	} else {
		arr := make([]int, last-first+1)
		for i := range arr {
			arr[i] = s.length - len(s.substringArray[first+i])
		}

		sort.Ints(arr)

		return arr
	}
}
